[Website] Add ADBC blog post

lidavidm opened the pull request.

lidavidm commented:
Pending ML vote and such

lidavidm commented:
Rendered: https://dynamic-jalebi-94dd65.netlify.app/blog/2022/10/03/arrow-adbc/

vibhatha commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,222 @@
+
+## Examples
+
+*Note: implementations are still under development, separate from the API standard itself. Examples are subject to change.*
```
Maybe?

```suggestion
*Note: implementations are still under development, separate from the API standard itself. Examples are subjected to change.*
```

or 

```suggestion
*Note: implementations are still under development, separate from the API standard itself. Examples can be subjected to change.*
```

lidavidm replied:
It's "subject to change".

lwhite1 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,222 @@
+</figure>
+
+**ADBC doesn't intend to replace JDBC or ODBC in general**, but we think ADBC makes more sense for applications that just want bulk columnar data access.
+
```
This seems almost like the benefits statement: Eg. "ADBC is an alternative to JDBC or ODBC for applications that need efficient bulk columnar data access...." Or even, "...an efficient, columnar alternative to JDBC/ODBC for analytical (data science?) applications."

lwhite1 reviewed (approved):
Very nice. I expect this will be an important milestone for Arrow.

lidavidm commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,222 @@
+  <img src="{{ site.baseurl }}/img/ADBCQuadrants.svg"
+       alt="ADBC is Arrow-native and database-agnostic, while Flight SQL is Arrow-native and database-specific. JDBC and ODBC are row-oriented and database-agnostic, and wire protocols like Postgres's and SQL Server's are row-oriented and database-specific."
+       width="90%" class="img-responsive">
+</figure>
```
Good point. Since we're talking about databases, how about a table instead? 
![image](https://user-images.githubusercontent.com/327919/207962602-98df7b4e-babc-466b-97c6-4a06651363b3.png)

lidavidm commented:
Thanks a lot for the feedback @lwhite1! I took your suggestions and also tried rewriting the intro. Hopefully this reads a little better.

lidavidm commented:
Preview updated (same link as before): https://dynamic-jalebi-94dd65.netlify.app/blog/2022/12/31/arrow-adbc/

lidavidm commented:
@ianmcook good idea, updated

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,251 @@
+
+ADBC combines the advantages of the latter two solutions under one API.
+In other words, ADBC provides a set of API definitions that client applications code to.
+These API definitions are Arrow based.
```
nit
```suggestion
These API definitions are Arrow-based.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,251 @@
+ADBC combines the advantages of the latter two solutions under one API.
+In other words, ADBC provides a set of API definitions that client applications code to.
+These API definitions are Arrow based.
+The application then links to or loads drivers for the actual database, that implement the API definitions.
```
```suggestion
The application then links to or loads drivers for the actual database, which implement the API definitions.
```

ianmcook commented on img/ADBCFlow2.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:lucid="lucid" width="800" height="600"><g transform="translate(0 0)" lucid:page-tab-id="fKyXq-xLtJEi"><path d="M0 0h800v600H0z" fill="#fff"/><path d="M262.5 6a6 6 0 0 1 6-6h88a6 6 0 0 1 6 6v588a6 6 0 0 1-6 6h-88a6 6 0 0 1-6-6z" stroke="#3a414a" fill="#fff"/><use xlink:href="#a" transform="matrix(1,0,0,1,267.5 ...
```
I think it would be best to add bidirectional arrow heads to the lines connecting **ADBC API** to **Flight SQL ADBC Driver** and **libpq ADBC Driver**

lidavidm replied:
Updated the diagram, thanks!

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,251 @@
+  JDBC is row-oriented, and ODBC has implementation caveats, as discussed, that make it hard to use with Arrow.
+- **Vendor-agnostic**: ADBC drivers can implement the API using any underlying protocol, while Flight SQL requires server-side support that may not be easy to add.
+
+ADBC is both vendor-agnostic and columnar, while Flight SQL is vendor-specific, and JDBC/ODBC are row-oriented:
```
This seems to contract the prior assertion that "ODBC can support columnar data"

lidavidm replied:
Yeah, I'm mostly trying to characterize ODBC as "row-oriented, with some columnar capabilities", I'll tweak this

lidavidm replied:
Updated.

Also, tweaked wording below the table to emphasize the API vs protocol distinction behind ADBC/Flight SQL.

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,251 @@
+    </tr>
+    <tr>
+      <th scope="row">Row-oriented</th>
+      <td>JDBC<br>ODBC</td>
```
```suggestion
      <td>JDBC<br>ODBC (typically)</td>
```

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+
+As mentioned, ADBC is an API standard.
+You can read the [C/C++ header][adbc.h], [Go interfaces][adbc-go], and the [Java interfaces][adbc-core] for yourself.
+The project has built several libraries around this standard:
```
How can the project be accessed, is this supposed to be the repo, or...?

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+
+- Driver implementations implement the API standard on top of a particular protocol or vendor.
+  See the drivers for [libpq (PostgreSQL)][adbc-libpq] and [JDBC][adbc-jdbc].
+- The driver manager library provides an easy way to use multiple drivers from a single program, and also load drivers at runtime if desired.
```
This library sounds incredible -- where can we get it? Should it be mentioned earlier? It sounds like a very impressive feature to me.

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+  JDBC is row-oriented, and ODBC has implementation caveats, as discussed, that make it hard to use with Arrow.
+- **Vendor-agnostic**: ADBC drivers can implement the API using any underlying protocol, while Flight SQL requires server-side support that may not be easy to add.
+
+ADBC is both vendor-agnostic and columnar, while Flight SQL is vendor-specific, and JDBC/ODBC are (generally) row-oriented:
```
```suggestion
Where ADBC fulfills both of these needs, Flight SQL is vendor-specific, and JDBC/ODBC are (generally) row-oriented:
```
This feels a bit cleaner to me, but I also notice that it just restates the bullets above -- maybe it should match the language on the JDBC/ODBC part to directly mention Arrow, or this sentence could be dropped?

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+  </tbody>
+</table>
+
+**ADBC doesn't intend to replace JDBC or ODBC in general**, but we think ADBC makes more sense for applications that just want bulk columnar data access.
```
Does it make more sense? Why would an application want this? Would the reader already have all this in mind?

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+       width="90%" class="img-responsive">
+</figure>
+
+To start using ADBC today, see the [repository][adbc] on how to build and install the packages, or look at the [documentation][adbc-docs] for a short tutorial.
```
Does the documentation include build and install instructions? If so, would it be better to point people to the repo for the code, then the docs for how to build, install, and use?

ksuarez1423 reviewed (commented):
There's a lot of really good content here, I just think it's holding back a bit on positivity, and I think needs more clarity on who it targets -- sometimes, the article introduces basic topics with explanation, then comes out of left field with one that seems to be assumed prior knowledge. It's not clear to me why it does what when.

lidavidm commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+- Non-columnar protocols, like the PostgreSQL wire format;
+- Non-columnar API abstractions, like JDBC/ODBC.
+
+In other words: **ADBC is a single API for getting Arrow data in and out of databases**.
```
Right, three classes within a single API. I'll think about rewording this a bit.

lidavidm commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+- [Nanoarrow][nanoarrow] is used to implement some ADBC drivers, making them lightweight and independent of the core Arrow libraries.
+
+Also, ADBC provides API support for the emerging [Substrait][substrait] specification.
+Applications can pass queries either as strings (of SQL) or binary blobs (of serialized Substrait plans).
```
I think some people will conversely nit me and say that really the driver just gets text/bytes and it doesn't _have_ to be SQL or Substrait but probably that will be a less common point

ksuarez1423 replied:
Will it function with anything other than SQL or Substrait?

lidavidm reviewed (commented):
Thanks for the thorough review. I tried to address all points. I ended up rewriting most of the post. The preview is still here: https://dynamic-jalebi-94dd65.netlify.app/blog/2022/12/31/arrow-adbc/

lidavidm commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,252 @@
+
+Similarly, within the Arrow project, ADBC does not replace Flight SQL, but instead *complements* it.
+ADBC gives client applications a uniform **API** to multiple database protocols and APIs—including Flight SQL.
+Meanwhile, Flight SQL gives database servers a single **wire protocol** that can be implemented to simultaneously target ADBC users, [JDBC users][flight-sql-jdbc], and ODBC users.
```
We could follow up with a short post about these two specifically, though, since I expect it'll be a common question. (Or just update the FAQ.)

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+  These standards are here to stay, and it makes sense for databases to support them for applications that want them.
+  But when both the database and the application are columnar, that means converting data into rows for JDBC/ODBC, only for the client to convert them right back into columns!
+  Performance suffers, and developers have to spend time implementing the conversions.
+- *Use JDBC/ODBC to Arrow conversion libraries*.
```
reads more clearly with hyphens
```suggestion
- *Use JDBC/ODBC-to-Arrow conversion libraries*.
```

lidavidm replied:
thanks!

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+  <img src="{{ site.baseurl }}/img/ADBCFlow2.svg" alt="A diagram showing the query execution flow with ADBC." width="90%" class="img-responsive">
+  <figcaption>The query execution flow with two different ADBC drivers.</figcaption>
+</figure>
+
```
Here the numbered list follows the diagram, whereas above, the numbered list precedes the diagram. It'd probably be best to have the order the same in both places.

lidavidm replied:
Updated (diagram precedes list in both places)

ksuarez1423 commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+  (Look at all the [connectors](https://trino.io/docs/current/connector.html) that Trino implements.)
+  And databases like PostgreSQL don't offer an option supporting Arrow in the first place.
+
+So in the status quo, clients must choose between either tedious integration work or leaving performance on the table.
```
```suggestion
As is, clients must choose between either tedious integration work or leaving performance on the table. We can make this better.
```

I think this could be punchier as a section ender.

lidavidm replied:
Thanks! Updated the preview.

lidavidm commented:
@ksuarez1423, @ianmcook any final comments? (Especially since I rewrote the post quite heavily.)

Thanks for all your help & happy New Year's!

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+2. The query is passed on to the driver.
+3. The driver translates the query to a database-specific protocol and sends it to the database.
+4. The database executes the query and returns the result set in a database-specific format.
+5. The driver translates the result format into the JDBC/ODBC API.
```
Is this what this means?
```suggestion
5. The driver translates the result into the format required by the JDBC/ODBC API.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+
+When columnar data comes into play, however, problems arise.
+JDBC is a row-oriented API, and while ODBC can support columnar data, the type system and data representation is not a perfect match with Arrow.
+So generally, columnar data must be converted to rows between steps 5 and 6, spending resources without performing "useful" work.
```
```suggestion
So generally, columnar data must be converted to rows in step 5, spending resources without performing "useful" work.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+  But when both the database and the application are columnar, that means converting data into rows for JDBC/ODBC, only for the client to convert them right back into columns!
+  Performance suffers, and developers have to spend time implementing the conversions.
+- *Use JDBC/ODBC-to-Arrow conversion libraries*.
+  Libraries like [Turbodbc][turbodbc] and [arrow-jdbc][arrow-jdbc] handle row to columnar conversions for clients.
```
```suggestion
  Libraries like [Turbodbc][turbodbc] and [arrow-jdbc][arrow-jdbc] handle row-to-columnar conversions for clients.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+- *Use vendor-specific protocols*.
+  For some databases, applications can use a database-specific protocol or SDK to directly get Arrow data.
+  For example, applications could use Dremio via [Arrow Flight SQL][flight-sql].
+  But client applications that want to use multiple database vendors would need to integrate with each of them.
```
```suggestion
  But client applications that want to support multiple database vendors would need to integrate with each of them.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+## Introducing ADBC
+
+ADBC is an Arrow-based, vendor-netural API for interacting with databases.
+Applications that use ADBC just get Arrow data.
```
```suggestion
Applications that use ADBC receive Arrow data.
```

ianmcook replied:
Or "simply receive" if you want to emphasize that no conversion is necessary.

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+
+The application only deals with one API, and only works with Arrow data.
+
+For example, in Python, the ADBC packages offer a familiar [DBAPI 2.0 (PEP 249)][pep-249]-style interface, with extensions to get Arrow data.
```
This reference to the Python ADBC packages seems abrupt. Add a sentence before giving the context. Maybe like this:
```suggestion
ADBC API and driver implementations are under development. For example, in Python, the ADBC packages offer a familiar [DBAPI 2.0 (PEP 249)][pep-249]-style interface, with extensions to get Arrow data.
```

ianmcook commented on img/ADBC.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:lucid="lucid" width="800" height="600"><g transform="translate(0 0)" lucid:page-tab-id="0_0"><path d="M0 450h800v150H0zM0 0h800v150H0z" stroke="#000" stroke-opacity="0" stroke-width="2" fill="#f2f3f5"/><path d="M.5 451a.5.5 0 0 1 .5-.5h798a.5.5 0 0 1 .5.5.5.5 0 0 1-.5.5H1a.5.5 0 0 1-.5-.5zM.5 151a.5.5 0 0 1 . ...
```
This text in the diagram seems confusing:
>The database only works with Arrow data, regardless of the actual client.

The database does not necessarily "work with" Arrow data. It (ultimately) emits Arrow data, but internally might be working with the data in some other database-specific format. Is there a better way to express what you mean by this?

lidavidm replied:
I meant something like that, yeah. The database only has to implement one (columnar, Arrow-based) endpoint/protocol but can support Arrow-native and 'traditional' clients.

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+ADBC works as part of the Arrow ecosystem to "cover the bases" for database interaction:
+
+- Arrow offers a universal columnar data format,
+- Arrow Flight SQL offers a universal server database protocol,
```
```suggestion
- Arrow Flight SQL offers a universal wire protocol for database servers,
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+
+- Arrow offers a universal columnar data format,
+- Arrow Flight SQL offers a universal server database protocol,
+- and ADBC offers a universal client database API.
```
```suggestion
- and ADBC offers a universal API for database clients.
```

ianmcook commented on _posts/2022-12-31-arrow-adbc.md:
```diff
@@ -0,0 +1,217 @@
+
+ADBC was only possible with the help and involvement of several Arrow community members and projects.
+In particular, we would like to thank members of the [DuckDB project][duckdb] and the [R DBI project][dbi], who constructed prototypes based on early revisions of the standard and provided feedback on the design.
+And ADBC builds on existing Arrow projects, including the [Arrow C Data Interface][c-data-interface] and [Nanoarrow][nanoarrow].
```
We've generally been keeping "nanoarrow" all lowercase, like pandas or dplyr.
```suggestion
And ADBC builds on existing Arrow projects, including the [Arrow C Data Interface][c-data-interface] and [nanoarrow][nanoarrow].
```

lidavidm commented:
Updated, thanks Ian! I tweaked the diagram too.

Updated preview: https://dynamic-jalebi-94dd65.netlify.app/blog/2023/01/04/introducing-arrow-adbc/ (since I had to bump the date forward)

lidavidm commented:
I'll be publishing this in a bit. Thanks to all who reviewed!
